### PR TITLE
[Tooling] Identify the source library when merging strings and add tools:ignore="UnusedResources" when appropriate

### PR DESCRIPTION
**Pull request by AliSoftware**
🚧  We need to land this [PR in release-toolkit about `tools:ignore`](https://github.com/wordpress-mobile/release-toolkit/pull/354) first, then update the reference in the `Pluginfile`, before merging this PR (hence the current "Not Ready for Merge" label)

### Identify where the strings are merged from

This PR uses the new feature currently being introduced in https://github.com/wordpress-mobile/release-toolkit/pull/351 in order to identify which libs each merged string came from.

This will be useful:
 - For developers and platform engineers, to understand where the source of truth of each string resides
 - To better follow the source of each string, and potentially help us do some spring cleaning of unused strings in the future

This work started after internal discussion paaHJt-3fg-p2 about how strings from libs are merged into the apps' strings to be sent to translation.

### Add `tools:ignore="UnusedResources"` to merged strings from binary dependencies

When merging strings from binary dependencies, the linter is not capable of detecting that the strings that will get merged in the app's `strings.xml` are actually used in the binary dependencies and that the app will just override their value at runtime. Thus, we need to add `tools:ignore="UnusedResources"` to all the strings merged from binary dependencies to avoid extra linter warnings.

This uses the new feature of the `release-toolkit` implemented in https://github.com/wordpress-mobile/release-toolkit/pull/354

Note that for strings merged from local modules, we don't want to add `tools:ignore="UnusedResources"` for those, because [we have the `checkDependencies true` option enabled on our lint config of the WordPress project](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/build.gradle#L265), so the linter WILL be able to detect when a string merged into the app's `strings.xml` is in fact used by the code from the module.

Note that this doesn't always necessarily means that any string merged from a module but for which the linter would warn us about it being an unused resource… would be a true positive and genuinely unused string: such string could still be a string value that we decided to override in the module… from a transitive dependency that the module itself depends on, for example. But still, for those we do want the linter to emit the warnings, so that we can decide manually if it's genuinely unused and we should remove the string from the module's `strings.xml` (and the app's), or if it's a false positive and we should add `tools:ignore="UnusedResources"` manually to it, case by case.

## To Test

 - Run `bundle exec fastlane localize_libs`
 - Verify that the `strings.xml` file was not changed after the run (I've already run it on commit 65f00b3e6d70a2da67c4e470a9d8b78a26a7925f)
 - Manually change some entries in the `WordPress/src/main/res/values/strings.xml` file:
   - Remove some random entries which come from a library
   - Change the text of a random entry which come from a library
   - Change the `a8c-src-lib` attribute of some entries to a different dummy value
   - Remove the `a8c-src-lib` attribute of some entries
   - Remove a `tools:ignore="UnusedResources"` from a string which was merged from one of the binary dependencies (e.g. a string that came from the login lib)
   - Change a `tools:ignore="UnusedResources"` which came from a binary lib to `tools:ignore="X"`
   - Change a `tools:ignore="UnusedResources"` which came from a binary lib to `tools:ignore="X,UnusedResources,Y"`
 - Run `bundle exec fastlane localize_libs` again
 - Notice how all your changes got reverted back.
   - The entries you removed (or for which you removed attributes) should be added back, except at the end of the XML
   - The entries that you changed should be updated in-place to restore the previous values — the one from the lib
   - The entries for which you changed `UnusedResources` to `X` should now be `X,UnusedResources`, and the one with `X,UnsedResources,Y` should still have the same attribute
 - Run `./gradlew mergeWordPressVanillaReleaseResources` and verify that there is no error nor warning about the `values/strings.xml` file
   - Side note that you will still have many "Multiple substitutions specified in non-positional format" warnings on `values-{locale}/strings.xml` files, but those are unrelated and can be ignored (they will disappear once [the fixes on the originals](https://github.com/wordpress-mobile/WordPress-Android/pull/16177) finally get fixed by polyglots in GlotPress' translations too)

### CODE REVIEW

File: fastlane/Pluginfile
Comment (AliSoftware) [inline fastlane/Pluginfile]: We need to release a new version of the release-toolkit then update this reference back to the new version once [this](https://github.com/wordpress-mobile/release-toolkit/pull/351) lands.

File: WordPress/src/main/res/values/strings.xml
Comment (AliSoftware) [inline WordPress/src/main/res/values/strings.xml]: This new feature allows us to see that, interestingly, while `email_invalid ` comes from the Login library, `username_or_password_incorrect` for example does not (while I would have initially thought given context that such a string would be defined there too…) 🙃

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb]:
```suggestion
    { library: "Image Editor", strings_path: "./libs/image-editor/ImageEditor/src/main/res/values/strings.xml", source_id: 'module:image-editor' },
    { library: "WordPress Editor", strings_path: "./libs/editor/WordPressEditor/src/main/res/values/strings.xml", source_id: 'module:editor' }
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb]:
```suggestion
      strings_file_path: 'bundle/android/strings.xml',
      source_id: 'gutenberg'
```

**Review by mokagio — APPROVED** [review]
Testing in a dedicated branch. [My changes](https://github.com/wordpress-mobile/WordPress-Android/commit/2806451ffb4783ba2121a14933ebdc94983433ef) were [reverted](https://github.com/wordpress-mobile/WordPress-Android/commit/a27bd97795b6503ca7a150de8889771c0402f3d2) as described.

👍 

Good to go once the new toolkit version ships.

**Review by mokagio — COMMENTED** [review]
I re-requested a review from myself because I wanted to "comment" instead of "[approve](https://github.com/wordpress-mobile/WordPress-Android/pull/16280#pullrequestreview-937486123)," to avoid accidental merges (even though we're pretty safe between this being a draft and having the Do Not Merge label)

**Comment by AliSoftware** [comment]
Notes to self of what I want to do on this PR before merging it:
 - [x] Make all `source_id` values be explicit in the `REMOTE_LIBRARIES_STRINGS_PATHS`, similar to the [GitHub suggestion I made for the case of Gutenberg Mobile](https://github.com/wordpress-mobile/WordPress-Android/pull/16280#discussion_r845985693). Will allow us to use prefixes like `module:` vs `lib:` prefix or similar as well.
 - [x] Implement the feature in the release-toolkit to automatically add `tools:ignore="UnusedResources"` to the string nodes merged from libraries
    - This would be an optional parameter to pass to the action in the Hash describing each library to merge. By default it would not add it, but if that new param like `add_tools_ignore_attr: true` is set, it would add it.
    - We'd then only use that attribute for `REMOTE_LIBRARIES_STRINGS_PATHS` — since the linter can't detect strings that seems unused to it but are actually used by those binary libs — but would not use it for `LOCAL_LIBRARIES_STRINGS_PATHS` — because those libs are local dependencies, and we have the `checkDependencies: true` set in our `WordPress/build.gradle`, so the linter currently checks those (and any key that the linter would then see unused would be more likely to be a true positive there)
    - Then we'd test that new feature of "auto-add-tools-ignore" on this draft PR, and if it works successfully, we would then finally be able to create a new version of the `release-toolkit` that would contain both that `source_id` and the `add_tools_ignore_attr` features in the same release
 - [x] Once this new release of the toolkit is out, point the `Pluginfile` back to a version number so we can finally merge

**Comment by AliSoftware** [comment]
@mokagio I've made a couple of changes:
 - Rebased the PR on top of `trunk`
 - Made the `source_id` identifiers explicit in the fastfile for each library being merged
 - Also and more importantly, I've now made use of [the new feature adding `tools:ignore` automatically](https://github.com/wordpress-mobile/release-toolkit/pull/354) that I recently added to the release-toolkit — after discussing the idea with @ParaskP7 recently over chat.

In practice, all the strings that are currently in our app's `strings.xml` already have the `tools:ignore="UnusedResources"` attribute, so that last change would not really change anything in the current state (in fact, running `fastlane localize_libraries` with that option on the current `strings.xml` from trunk doesn't make any change to it). But that would be useful for future strings added to those libraries in the future.

**Comment by AliSoftware** [comment]
PR is now ready for review, but would require the following before merging:
 - [x] Update the `fastlane/Pluginfile` to point to a stable version of the `release-toolkit` once PRs are merged
 - [x] Maybe undo the changes made to the `strings.xml` file in this PR, to only let those changes be done during code freeze on next cycle — otherwise I can easily see the changes made in this PR quickly diverging from changes made in `trunk` when we'll decide to merge, and potentially creating conflicts or bad merges of the `strings.xml` file

**Comment by mokagio** [comment]
> Maybe undo the changes made to the `strings.xml` file in this PR [to avoid divergence]

Seems wise.

**Comment by AliSoftware** [comment]
@mokagio I've now updated the `Pluginfile` to point to the new release-toolkit `4.1.0`, and rebased the commits to remove the noise and the discard the changes made to the `strings.xml` by this PR. This is now ready for review.

PS: I know you've already approved the PR (while it was still in Draft state), so I could actually merge it right away, but just wanted to give you a chance to do a last check before hitting the green button 🙂 

### To Test

 - Create a testing branch
 - Remove a couple of string tags from the `WordPress/src/main/res/values/strings.xml` file, change the copy of some others that you clearly know came from a library, and remove the `tools:ignore` attribute of other lib ones too
 - Run `bundle exec fastlane localize_libraries`
 - Check that the lane succeeds and creates a commit changing the `strings.xml` file
 - Check that the changes made in that `strings.xml` file are consistent with what we expect, especially:
   - re-introducing, at the end, the XML nodes you removed but were from a lib
   - updating in-place the XML nodes for which you change the copy/value and were from a lib
   - adding `a8c-src-lib` attribute to all the strings nodes which came from libs
   - re-adding `tools:ignore` to the XML nodes which came from a lib but for which you deleted the attribute

### Announcement

I've made a draft for a P2 post announcement for this change and shared the link to you in Slack.

Feel free to review it and adjust as you see fit, then hit publish when you think the post is ready (assuming you approved and merged this PR since)

**Review by mokagio — APPROVED** [review]
Tested by changing random stuff via https://github.com/wordpress-mobile/WordPress-Android/pull/16399.

All behaving as expected, as far as I could see. 👍

**Comment by mokagio** [comment]
Alos, post published. Internal ref: paqN3M-xk-p2

**Pull request merged**
